### PR TITLE
[bugfix] reduce float value error when adding noise

### PR DESCRIPTION
**Pull request by gameofdimension**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by bghira** [comment]
cc @sayakpaul @linoytsaban @yiyixuxu

**Comment by gameofdimension** [comment]
it actually adds no noise at all for step 0/2 when using torch.bfloat16. run the following code to reproduce that.

a demo output

image

```
import torch
from diffusers import DDPMScheduler


def demo(step, latents, noise_scheduler):
    bsz = latents.shape[0]
    noise = torch.randn_like(latents)
    timesteps = torch.randint(
        step, step+1,
        (bsz,), device=latents.device)
    timesteps = timesteps.long()

    noisy_latents = noise_scheduler.add_noise(
        latents, noise, timesteps)

    delta = (noisy_latents-latents).abs().max()
    return delta.item()


def main():
    pretrained_model_name_or_path = '/apps/dat/file/llm/model/stable-diffusion-v1-5'
    noise_scheduler = DDPMScheduler.from_pretrained(
        pretrained_model_name_or_path, subfolder="scheduler")
    bsz = 2
    latents = torch.randn(
        (bsz, 4, 64, 64),
        dtype=torch.bfloat16,
        device='cuda')

    finfo = torch.finfo()
    step = 0
    delta = demo(step, latents, noise_scheduler)
    print(f"delta after add step {step} noise", delta)
    assert delta < finfo.eps

    step = 1
    delta = demo(step, latents, noise_scheduler)
    print(f"delta after add step {step} noise", delta)
    assert delta < finfo.eps

    step = 2
    delta = demo(step, latents, noise_scheduler)
    print(f"delta after add step {step} noise", delta)


if __name__ == '__main__':
    main()
```

**Comment by bghira** [comment]
well, step 0 is timestep 0. and there's an off-by-one error throughout Diffusers because of how the scheduler steps in order with the rest of the steps. so you get to the end of the schedule and there's one prediction missing. it has to add an extra zero sigma step to complete it. especially with DDIM.

```py
Timesteps: tensor([0], device='mps:0')
delta after add step 0 noise 0.0
Timesteps: tensor([1000], device='mps:0')
delta after add step 1000 noise 5.6875
```

is that what you're observing? it's normal not to add noise when the sigma is zero.

**Comment by bghira** [comment]
ah i see. using float32 makes the first value 0.13. but the batch size changes the amount of noise too.

**Comment by gameofdimension** [comment]
> well, step 0 is timestep 0. and there's an off-by-one error throughout Diffusers because of how the scheduler steps in order with the rest of the steps. so you get to the end of the schedule and there's one prediction missing. it has to add an extra zero sigma step to complete it. especially with DDIM.
> 
> ```python
> Timesteps: tensor([0], device='mps:0')
> delta after add step 0 noise 0.0
> Timesteps: tensor([1000], device='mps:0')
> delta after add step 1000 noise 5.6875
> ```
> 
> is that what you're observing? it's normal not to add noise when the sigma is zero.

what about step 1

**Comment by gameofdimension** [comment]
> ah i see. using float32 makes the first value 0.13. but the batch size changes the amount of noise too.

i don't think so. batch size is irrelevant.

**Comment by bghira** [comment]
```py
    def add_noise(
        self,
        original_samples: torch.Tensor,
        noise: torch.Tensor,
        timesteps: torch.IntTensor,
    ) -> torch.Tensor:
        if original_samples.dtype == torch.bfloat16::
            original_samples = original_samples.to(dtype=torch.float32)
        if noise.dtype == torch.bfloat16:
            noise = noise.to(original_samples.device, dtype=torch.float32)
```

update DDPM/DDIMScheduler to have this

**Comment by bghira** [comment]
we're already handling casting on mps using UniPC scheduler but DDPM or DDIM doesn't seem to do it. probably others with the issue

**Comment by bghira** [comment]
```
    noise_scheduler = EulerDiscreteScheduler.from_pretrained(
        pretrained_model_name_or_path, subfolder="scheduler", rescale_betas_zero_snr=True, timestep_spacing="trailing")
    bsz = 1
    print(f"Batch size: {1}")
    latents = torch.randn(
        (bsz, 4, 64, 64),
        dtype=torch.bfloat16,
        device='mps')

```


```
Batch size: 1
Timesteps: tensor([0], device='mps:0')
delta after add step 0 noise 0.115234375
```

Euler works on bf16 noise/latents

**Comment by gameofdimension** [comment]
> we're already handling casting on mps using UniPC scheduler but DDPM or DDIM doesn't seem to do it. probably others with the issue

it seems that DDPM is usually used when do training.

**Comment by bghira** [comment]
huggingface uses euler for training

**Comment by bghira** [comment]
also this is an inference/training issue not necessarily just for training.. though it really only applies during img2img then and i'm not sure what the implications are. it's definitely worse for training.

**Comment by sayakpaul** [comment]
Cc: @yiyixuxu could you give this a look?

**Comment by bghira** [comment]
again cc @sayakpaul @linoytsaban @yiyixuxu

**Comment by sayakpaul** [comment]
Sorry for the delay on my part, @gameofdimension! Did you notice the same behaviour for other scripts or is this specific to ControlNet?

**Comment by gameofdimension** [comment]
it seems like all calls of `noise_scheduler.add_noise` are susceptible to this issue

**Comment by yiyixuxu** [comment]
let's fix this for training first 
if we run into issues with inference then we can look into scheduler

**Comment by gameofdimension** [comment]
> let's fix this for training first if we run into issues with inference then we can look into scheduler

what else should i do?

**Comment by bghira** [comment]
idk honestly its ready for merge. diffusers team are you all ok?

**Review by yiyixuxu — APPROVED** [review]
